plugins/catalog-backend-module-github/src/providers/GithubDiscoveryEntityProvider.ts

```diff
@@ -0,0 +1,297 @@
+  }
+
+  private createScheduleFn(schedule: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return schedule.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const gitHubConfig = this.integrations.github.byUrl(
+      this.providerConfig.orgUrl,
+    )?.config;
+
+    if (!gitHubConfig) {
+      throw new Error(
+        `There is no GitHub integration that matches ${this.providerConfig.orgUrl}. Please add a configuration entry for it under integrations.github`,
+      );
+    }
+
+    const { org, repoSearchPath } = parseUrl(this.providerConfig.orgUrl);
+
+    const { headers } = await this.githubCredentialsProvider.getCredentials({
+      url: this.providerConfig.orgUrl,
+    });
+
+    const client = graphql.defaults({
+      baseUrl: gitHubConfig.apiBaseUrl,
+      headers,
+    });
+
+    const startTimestamp = Date.now();
+
+    this.logger.info(
+      `Reading GitHub repositories from ${this.providerConfig.orgUrl}`,
+    );
+
+    const expectedEntityFiles = this.providerConfig.files;
+
+    const { repositories } = await getOrganizationEntities(
+      client,
+      org,
+      expectedEntityFiles,
+    );
+
+    const matching = repositories.filter(
+      r => !r.isArchived && repoSearchPath.test(r.name),
+    );
+
+    const duration = ((Date.now() - startTimestamp) / 1000).toFixed(1);
+
+    this.logger.info(
+      `Read ${repositories.length} GitHub repositories (${matching.length} matching the pattern) in ${duration} seconds`,
+    );
+
+    const fileKeys = expectedEntityFiles.map(fileConfig =>
+      camelCase(fileConfig.file),
+    );
+
+    const repoEntities: Array<YamlObject | Array<YamlObject>> = [];
+    for (const entity of matching) {
+      for (const key of fileKeys) {
+        const entityObj: any = entity[key as keyof typeof entity];
+        if (entityObj) {
+          const yamlEntities = yamlStringToObjects(entityObj?.text);
+          const validEntities = yamlEntities.filter(v => validateYamlObj(v));
+          repoEntities.push(validEntities);
+        }
+      }
+    }
+
+    const validEntities = repoEntities.flat();
+
+    this.logger.info(
+      `Adding ${validEntities.length} valid entities to the catalog`,
+    );
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: validEntities.map(e => ({
+        locationKey: `github-discovery-graphql-provider:${this.providerConfig.id}`,
+        entity: withLocations(this.providerConfig.orgUrl, org, e),
+      })),
+    });
```

Comment:
Discovery processors should generally emit location entities that point to the targets, rather than parsing the targets directly. See for example https://github.com/backstage/backstage/blob/aba7585b6a6e3a11c68aba10645f037345e3acfc/plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts#L173-L179

This design makes sure we both have location entities in the ancestry chain that helps us highlight errors as we otherwise have no entities to attach parsing errors to. It also defers the parsing and scanning of the in-repo location to processors, allowing integrators to plug in their own readers and parsers without deep modification of the discovery providers.

In the end the only thing we really want to do is to lift over the logic from the [GitHubDiscoveryProcessor](https://github.com/backstage/backstage/blob/aba7585b6a6e3a11c68aba10645f037345e3acfc/plugins/catalog-backend-module-github/src/GithubDiscoveryProcessor.ts#L115-L156)